securedrop/upload-screenshots.py

```diff
@@ -0,0 +1,218 @@
+#!/usr/bin/env python3
+
+from glob import glob
+from urllib.parse import urljoin
+
+import os
+import argparse
+import re
+import requests
+import sys
+
+# Used to generate URLs for API endpoints and links; exposed as argument
+DEFAULT_BASE_URL = "https://weblate.securedrop.org"
+
+# Where we look for screenshots: the page layout test results in English
+SCREENSHOTS_DIRECTORY = os.path.abspath(
+    os.path.join(os.path.dirname(__file__), "tests/pageslayout/screenshots/en_US")
+)
+
+# What pattern we expect them to match
+SCREENSHOTS_GLOB = "*.png"
+
+# Regular expression rules that are applied sequentially to transform a
+# filename into the canonical title we give that screenshot in Weblate.
+#
+# Example conversion: "source-session_timeout.png" -> "source: session timeout"
+CANONICALIZATION_RULES = [(r"\.png$", ""), (r"-", ": "), (r"_", " ")]
+
+# Weblate organizes internationalization work into projects and components,
+# which are part of many URLs, and need to be referenced in some API requests.
+PROJECT_SLUG = "securedrop"
+COMPONENT_SLUG = "securedrop"
+
+# Request limit for the number of pagination requests to issue before aborting.
+REQUEST_LIMIT = 50
+
+
+def main():
+    """
+    Uses the generic WeblateUploader class below to run a SecureDrop screenshot
+    upload.
+    """
+    parser = argparse.ArgumentParser(
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter
+    )
+    parser.add_argument(
+        "--token",
+        help="API token for accessing the weblate API",
+        type=str,
+        required=True,
+    )
+    parser.add_argument(
+        "--baseURL",
+        help="Weblate base URL",
+        default=DEFAULT_BASE_URL,
+        type=str,
+        required=False,
+    )
+    args = parser.parse_args()
+
+    screenshot_files = glob(os.path.join(SCREENSHOTS_DIRECTORY, SCREENSHOTS_GLOB))
+    if len(screenshot_files) == 0:
+        print(
+            "Page layout test results not found. Run this command from the SecureDrop"
+        )
+        print("base directory to generate the English language screenshots:\n")
+        print("  LOCALES=en_US make translation-test")
+        print("\nThis will take several minutes to complete.")
+        sys.exit(1)
+
+    uploader = WeblateUploader(
+        token=args.token,
+        base_url=args.baseURL,
+        project=PROJECT_SLUG,
+        component=COMPONENT_SLUG,
+        files=screenshot_files,
+        request_limit=REQUEST_LIMIT,
+        canonicalization_rules=CANONICALIZATION_RULES,
+    )
+    uploader.safe_upload()
+
+
+class WeblateUploader(object):
+    """
+    Manages Weblate screenshot batch uploads, matching filenames against
+    titles of existing screenshots to create/update as appropriate.
+    """
+
+    def __init__(
+        self,
+        token,
+        base_url,
+        project,
+        component,
+        files,
+        request_limit,
+        canonicalization_rules=[],
```

Comment:
I don't think so either, but it's reasonable to avoid. These rules are intended to not be mutable, so switched to a tuple in 15ef8cb3e8bd3b5ce2f41e1f95a27754f15c4ae8